File: Programmation réseaux/exo5.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <sys/socket.h>
#include <netinet/in.h>
#include <unistd.h>

void relay(int de, int vers) 
{
    char buffer[1024] = {0};
    dprintf(vers, "L'autre client écrit...\n");
    dprintf(de, "Saisissez la réponse : ");
    int n = read(de, buffer, 1024);
    if (n > 0) 
    {
        dprintf(vers, "L'autre client dit : %s", buffer);
    }
}

int main() 
{
    int serv = socket(AF_INET, SOCK_STREAM, 0);
    struct sockaddr_in addr = {AF_INET, htons(8080), INADDR_ANY};
    bind(serv, (struct sockaddr*)&addr, sizeof(addr));
    listen(serv, 2);

    printf("En attente de deux clients...\n");
    int c1 = accept(serv, NULL, NULL);
    printf("Client 1 connecté. En attente du second...\n");
    int c2 = accept(serv, NULL, NULL);
    printf("Match ! Début de la conversation.\n");

    dprintf(c1, "Connecté ! Vous commencez.\n");
    dprintf(c2, "Connecté ! Attendez le message de l'autre...\n");

    while (1) 
    {
        relay(c1, c2);
        relay(c2, c1);
    }

    return 0;
}

File: Programmation réseaux/exo5.py
import socket

def serve_chat():
    serv = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
    serv.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEADDR, 1)
    serv.bind(('0.0.0.0', 8080))
    serv.listen(2)

    print("En attente des clients...")
    c1, _ = serv.accept()
    print("Client 1 ok...")
    c2, _ = serv.accept()
    print("Client 2 ok. Relay actif.")

    c1.send(b"Connecte ! A vous l'honneur.\n")
    c2.send(b"Connecte ! Attendez le message...\n")

    try:
        while True:
            # Tour du Client 1
            c1.send(b"Saisissez la reponse : ")
            msg = c1.recv(1024)
            if not msg: break
            c2.send(b"L'autre client dit : " + msg)

            # Tour du Client 2
            c2.send(b"Saisissez la reponse : ")
            msg = c2.recv(1024)
            if not msg: break
            c1.send(b"L'autre client dit : " + msg)
    except:
        pass
    finally:
        c1.close()
        c2.close()

if __name__ == "__main__":
    serve_chat()